server/src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -202,17 +226,68 @@ private Criteria createModifiesDateFilteringCriteria(Consumer consumer, Date sta
 
         criteria = criteria.createCriteria("pool")
                 .add(Restrictions.or(
-                    // Dates overlap if the start or end date is in our range
-                    Restrictions.or(
-                        Restrictions.between("startDate", startDate, endDate),
-                        Restrictions.between("endDate", startDate, endDate)),
-                    Restrictions.and(
-                        // The dates overlap if our range is completely encapsulated
+                        // Dates overlap if the start or end date is in our
+                        // range
+                        Restrictions.or(Restrictions.between("startDate",
+                                startDate, endDate),
+                                Restrictions.between("endDate", startDate, endDate)),
+                Restrictions.and(
+                        // The dates overlap if our range is completely
+                        // encapsulated
                         Restrictions.le("startDate", startDate),
                         Restrictions.ge("endDate", endDate))));
         return criteria;
     }
 
+    /**
+     * A version of list Modifying that finds modifier products for a Set of
+     * entitlements.
+     * When dealing with large amount of entitlements for which it is necessary
+     * to determine their modifier products.
+     * @param entitlement
+     * @return Entitlements that are being modified by the input entitlements
+     */
+    public Set<Entitlement> batchListModifying(List<Entitlement> entitlements) {
+        Set<Entitlement> modifying = new HashSet<Entitlement>();
+
+        // Get the map of product Ids to the set of
+        // overlapping entitlements that provide them
+        ProductEntitlements pidEnts = getOverlappingForModifying(entitlements);
+        if (pidEnts.isEmpty()) {
+            // Empty collections break hibernate queries
+            return modifying;
+        }
+        // Retrieve all products at once from the adapter
+        List<Product> overlappingProducts =
+                productAdapter.getProductsByIds(pidEnts.getAllProductIds());
+        for (Product overlappingProduct : overlappingProducts) {
+            boolean modifies = false;
+            for (Entitlement inputEntitlement : entitlements) {
+                if (overlappingProduct.modifies(inputEntitlement.getProductId())) {
+                    modifies = true;
+                    break;
```

Comment:
This seems incorrect to me as well. I think this needs a bit more test coverage.